### DISCUSSION TITLE
Where to register extension attributes that don't belong to a specific component?

### DISCUSSION THREAD

**Discussion by graue70** [explosion/spaCy#7425] [Help: Best practices]
## Setup

I have two extension attributes:
```python
Doc.set_extension("timing", default={})
Doc.set_extension("timing_total", getter=lambda doc: sum(doc._.timing.values()))
```

I have multiple custom components. I measure and store the time each of them takes in their respective `__call__()` methods like this:
```python
def __call__(self, doc: Doc) -> Doc:
    start_time = time.time()
    # Do something to doc
    doc._.timing[self.name] = time.time() - start_time
    return doc
```
After running the pipeline, this gives me the timing information:
```python
print(doc._.timing)
{"first_component": 1.63, "second_component": 4.91}
print(doc._.timing_total)
6.54
```

## Question

**In the docs, extension attributes are registered in the constructor of the component that sets them. My extension attributes are not related to a specific component, but rather to the pipeline/language as a whole. Every component writes to the extension attribute. Where do I set it?**

## Possible answers

1. Repeating the `Doc.set_extension()` in every component seems redundant and error prone. I would also have to make sure that the name is the same everywhere. If the getter was more complicated and wouldn't work as a lambda, I would even have to define the same function in every component.
2. Calling some code outside of the pipeline like this
    ```python
    nlp = spacy.from_config(config)
    set_global_extensions()
    doc = nlp("Some text")
    ```
    makes it hard to ensure it is always called (even in unit tests).
3. Adding a custom component which only sets the extension attributes seems kind of weird. If I call `nlp.pipe_names`, it shows a component which does nothing in the pipeline.

I basically need something like an `initialize` function on a pipeline/language which is not only called for training but also when the pipeline is loaded.

*(I originally asked three questions at once. I moved the other ones to https://github.com/explosion/spaCy/discussions/7445.)*

**Comment by ines** [comment]
> Calling some code outside of the pipeline like this 

I'd say this is the best solution for a use case like this. You could include this in the `before_creation`/`after_creation`/`after_pipeline_creation` callback via your config so it runs ince when the `nlp` object is created: https://spacy.io/api/data-formats#config-nlp

> Is my only choice to wrap the spaCy components in custom components like this?

An alternative option would be to insert an instance of a simple timing logging component in between all existing component (e.g. on `after_pipeline_creation`). This works well for debugging purposes, but it's maybe not that pretty because you end up with a bunch of extra components.

> Would adding the possibility to record timing for all steps of a pipeline be a welcome addition to the spaCy code? 

Just making it about timing is too specific and not really customisable enough – but I do see the general appeal of being able to hook into `Language.__call__` before/after a pipeline component is applied. The problem is that conceptually, this would actually end up being quite similar to the "components between components" setup, only much less explicit. The callback would have to receive the `Doc` object, but it shouldn't be modifying it – otherwise, the result would be very confusing and the callback would essentially just be a hidden pipeline component.

**Reply by graue70** [reply]
> I'd say this is the best solution for a use case like this. You could include this in the `before_creation`/`after_creation`/`after_pipeline_creation` callback via your config so it runs once when the `nlp` object is created: https://spacy.io/api/data-formats#config-nlp

I think this is exactly what I was looking for. Somehow I didn't find it in the docs. Thanks a lot.

Follow-up question: Is there something related in code or do the hooks only live in config files? If I load a config containing a `after_creation` hook with `nlp.from_config()` and do `nlp.to_config()`, will the resulting config have the `after_creation` hook as well?

*(I'll open another question for the other questions and remove them from this one.)*
